Add a Concourse CI pipeline that checks if PR author == latest commit signature author

issyl0 opened the pull request:
- This relates to the Concourse GitHub Resource PR (#378), and aims to reduce
  toil for us now we've changed what constitutes a mergeable PR that
  will successfully flow through the pipeline once merged.
- Humans are fallible and won't always remember to manually check that
  the PR author was the person who signed the latest commit.
- This makes it so that on every commit, Concourse will tell us. :muscle:

issyl0 committed:
On PRs, check if PR author == latest commit signature author

- This relates to the Concourse GitHub Resource PR, and aims to reduce
  toil for us now we've changed what constitutes a mergeable PR that
  will successfully flow through the pipeline once merged.
- Humans are fallible and won't always remember to manually check that
  the PR author was the person who signed the latest commit.
- This makes it so that on every commit, Concourse CI will check
  for us. :muscle:

issyl0 commented:
The shell script works locally. No idea about the Action here though (I'm nervous about the HEREDOC indentation) as it'll only trigger once it's on master.

I'll make this a full PR which we can discuss on Monday?

samcrang commented:
I tweaked this so that it doesn't tell you the person who opened the PR can approve it. This is enforced by the GitHub API but the output of this script doesn't help with that.

issyl0 commented:
This just checks the PR authorship, it doesn't check approvals, so this is fine.

issyl0 commented:
I made this be a Concourse pipeline instead of a GitHub Action. This way, we don't have to manage secrets/pipelines in two places.

samcrang commented:
@issyl0 that comment was meant to be on #378 🙈.

issyl0 commented:
This is working and good for review now.

Krenair commented on pipelines/ci/check-pr-author-matches-commit-signature.yml:
```diff
@@ -0,0 +1,114 @@
+- name: gsp-pr
+  type: pull-request
+  source:
+    access_token: ((pr-build-status))
```
pr-build-status should probably be named to indicate sensitivity. pr-build-status-secret or pr-build-status-access-token perhaps

issyl0 replied:
`pr-build-status-github-access-token` sounds good. Don't know how to edit the secret in SSM now that it exists, though. :thinking:

Krenair commented on pipelines/ci/check-pr-author-matches-commit-signature.yml:
```diff
@@ -0,0 +1,114 @@
+  type: pull-request
+  source:
+    access_token: ((pr-build-status))
+    disable_forks: true
```
this probably should not be set - people opening PRs from forks should also be told if their PR has an approval/signature problem, we don't want it to silently do nothing

Krenair commented on pipelines/ci/check-pr-author-matches-commit-signature.yml:
```diff
@@ -0,0 +1,114 @@
+- name: pull-request
+  type: docker-image
+  source:
+    repository: teliaoss/github-pr-resource
```
Discussed this, it's in use elsewhere in Verify so OK

Krenair commented on pipelines/ci/check-pr-author-matches-commit-signature.yml:
```diff
@@ -0,0 +1,114 @@
+        source:
+          repository: govsvc/task-toolbox
+      params:
+        GITHUB_TOKEN: ((pr-build-status))
```
needs rename as per above

issyl0 committed:
Run these checks on forks too

- People may be trusted developers, but they may not be in the right
  Verify/GSP teams for whatever reason in order to be able to push
  branches.
- When we notice PRs from forks, we should fix the permissions problems
  for GDS employees working on Verify, but we should also run these
  checks.

issyl0 committed:
Give the GitHub access token Concourse secret a better name

- I've edited this in AWS SSM Parameter Store too.

philandstuff reviewed (commented):
Rather than duplicating the github resource graphql query, is it possible to just use the new resource?

issyl0 commented:
Let's try it. When I wrote this, the resource wasn't released.

issyl0 commented:
Thinking about it: no, because the new resource does all the things (signatures and approvals), and approvals usually come after the last commit, so CI will turn red if the PR has no approvals even if it's half way implemented. What we want to check here is PR author signed the latest commit.
